### DISCUSSION TITLE
Proper text layout dependent on background-color

### DISCUSSION THREAD

**Discussion by TheEclectus** [mikke89/RmlUi#271] [Q&A] [answered]
Greetings. In stapling together my implementation of RmlUi's rendering interface, I've come across some bizarre functionality that I'm not sure is by design or a bug.

My test program consists of RmlUi 4.3 from vcpkg with a full Rml::RenderInterface implementation, and displays a simple .rml file:
```
<rml>
	<head>
		<title>Example</title>
		<link type="text/rcss" href="base.rcss"/>
		<style>
			p
			{
				display: block;
				margin: 0px 3px;
				background-color: #ccc;
				line-height: 1.0;
			}
			p.title
			{
				display: inline;
				background-color: #0cc;;
			}
			body
			{
				width: 500px;
				height: 200px;
				background-color: #ccc;
				font-family: Volter (Goldfish);
				font-size: 18px;
			}
			div#title
			{
				width: 100%;	
				background-color: #0cc;
				text-align: center;
			}
		</style>
	</head>
	<body>
		<handle move_target="#document">
			<div id="title">
				<p class="title">Title goes here, if there was one</p>
			</div>
		</handle>
		<p>Here are some</p>
		<p>p elements</p>
		<p>to test spacing</p>
	</body>
</rml>
```
`base.rcss` was copied directly from [https://mikke89.github.io/RmlUiDoc/pages/rml/html4_style_sheet.html](https://mikke89.github.io/RmlUiDoc/pages/rml/html4_style_sheet.html). 

![image](https://user-images.githubusercontent.com/66447088/150625415-3ccb6c0d-92ca-4710-9c2d-221531c6550f.png)

This is the expected output, which is accomplished with the above .rml and .rcss. However, when omitting the explicit `background-color` property from `p`, formatting is broken and it is apparently drawn at position `[0,0]` (as is corroborated by debugging my `RenderInterface` implementation, where `translation` becomes `{0,0}`):
![image](https://user-images.githubusercontent.com/66447088/150642726-bbb254e6-4541-4bf4-8f72-19b1bfa9497c.png)
This behavior is the same if a `background-color` with full transparency is specified.

Conversely, if `p.title` or `div#title` have no specified `background-color`, formatting remains in-tact. If `body` is without it, formatting is entirely broken and the text will be rendered in a blob in the top-left corner of the window: 
![image](https://user-images.githubusercontent.com/66447088/150643178-ca499945-0929-4f2a-a366-ec0eededbbc9.png)

Is this behavior intentional? Am I misinterpreting the sample code, which didn't seem to need a background-color explicitly specified to work? Any advice or guidance would be appreciated. Thank you!

**Comment by mikke89** [comment, marked as answer]
Hi, and welcome!

The first thing I will say is that in no way should the layout depend on the `background-color` property. I think there must be something else quirky going on here, I have never seen this behavior.

I tested your code and got the same layout as in your first screenshot. When I changed the background-color of the different rules you mention there is no change to the layout, as we would expect. Only the background changes.

Are you getting any log output / warnings? Please check that first.
Can you post a full example of a faulty render as well, so I can see if it reproduces here?

**Reply by TheEclectus** [reply]
Greetings!

It's definitely some vexing behavior. I don't believe that it's a rendering problem; I can still post the relevant code if need be (it's a very simplistic OpenGL renderer built with Magnum as an abstraction layer), but an offset of `{0.0, 0.0}` is being passed to my `RenderGeometry()` function (when the Document is in the top-left corner of the Context, anyways, The value being passed is always equal to the Document's offset). This seems to come from the result of `element->GetAbsoluteOffset(Box::BORDER)` being passed to `geometry.Render()` on line 52 of ElementBackgroundBorder.cpp. Exploring the memory of `element`, its `absolute_offset` variable is likewise `{0.0, 0.0}` at the time it's rendered. So unless I'm mistaken, there is something happening within RmlUi causing this behavior, and that it's not just a rendering bug.

I also feel I should mention I'm not touching Rml's state beyond the initial font loading and state setup, no programmatic layout alterations or anything.

**Reply by TheEclectus** [reply]
And now with my foot thoroughly lodged in my own mouth, I realize after stepping through Rml's routines that I'd neglected to multiply by the transform matrix if my `Rml::RenderInterface` implementation's `texture` parameter was non-null. Behavior is now entirely as is expected.

**Reply by mikke89** [reply]
Alright, I'm not sure I understand what's going on with the offset, but I'm happy it seems to be working for you now! Cheers.
